### ISSUE TITLE
Create New Model Training

### ISSUE THREAD

**Issue by Maddonix** [wg-lux/endoreg-db#96]
Acceptance Criteria:

1.  
- Add way to fetch annotations
- set selected frames' state to persist
- implement command to control extracted frames, only the ones flagged to persist should remain, others should be deleted
- implement way to fetch frames, should check if frame exists, if it doesnt we should extract it

2.
- Implement command to re-train model and create new model meta with bumped version

**Comment by Hamzaukw** [comment]
###  Summary of Current Implementation & Label Strategy

As of now, here’s how our dataset splitting and model training strategy is structured:

- We use **group-based splitting** by `old_examination_id` to ensure that all images from the same exam stay within the same split (train, val, or test). This prevents data leakage.

- We also tried a **random split**, which gave slightly better accuracy, but it introduces leakage — so we decided not to use it.

- With **group-wise splitting**, we noticed that it introduced significant label imbalance. For example, labels like `"grasper"`, `"snare"`, `"needle"`, and `"diverticule"` (all grouped as `"instrument"`) appear frequently, whereas labels like `"wound"` or `"blood"` are very rare. As a result, the model became biased toward more common classes.

To address this, we applied a **WeightedRandomSampler** during training:
-  Rare labels are oversampled — shown more often
-  Frequent labels are undersampled — shown less often

This improved model performance a bit not to much and helped the model learn rare classes better.

###  Model Architecture & Training

- We used **ImageNet pre-trained models** for feature extraction (EfficientNetB0, ResNet50).
- We also tried **EfficientNetV2M**, which gave better accuracy but required more training time (as compared to EfficientNetB0 ).
  - Example: ResNet50 took ~28 minutes for 8 epochs, EfficientNetB0 took ~15 minutes.

I also tried two loss functions:
- **Binary Cross-Entropy (BCE):** good for balanced labels
- **Focal Loss:** better for imbalanced labels, focuses more on hard-to-classify examples (better result)

I have gone through the Bucket/splitter approach and have some concerns, I would like to discuss

**What this approach does**: The Splitter class balances the number of samples across buckets, ensuring each bucket receives a roughly equal number of examples but it does not consider the label distribution within those examples.


**Problem:** If our dataset is highly imbalanced (e.g., if we have a small number of positive examples in a classification task), then uniform splitting across buckets may cause some buckets to have very few examples of the minority class.


**Reality:** This can lead to poor model performance because the model might train on data that's not representative of the real-world distribution of classes.


### Solution to problem - Hybrid Splitting Strategy (need to discuss)

Since our dataset will keep growing, i tried  a **hybrid splitting strategy** with persistent bucket mapping.

Here’s how it works:

1. Each group (`old_examination_id`) is randomly assigned to one of `NUM_BUCKETS` (e.g., 0–9).
2. This mapping is saved to disk (`bucket_map.pkl`) and reused — so the same exam always maps to the same bucket, even as the dataset grows.
3. From these buckets:
   - We **assign training buckets** based on config.
   - Then use **GroupShuffleSplit** to create validation and test splits.
   - During val/test splitting, we use `np.vstack()` + `MultiLabelBinarizer` to preserve **label distribution** (stratification).
   - Already using **WeightedRandomSampler**
4. Aim for:
   -  No exam-level leakage across splits
   - Balanced label distribution in val/test
   -  Scalability as new data arrives

 Note: `StratifiedKFold` doesn’t work here because it's designed for single-label tasks, not multi-label ones.

**Comment by Hamzaukw** [comment]
RECOMMENDATION FOR THOMAS (shared on nextcloud)


We can discuss this in detail next week but I was thinking of two bucket systems, test and train. 

First an image is, at least partly, annotated. At a certain point we decide to add an image to the fixed dataset.

At this point we do the following:
- check existing positive annotations of the image 
- check existing dataset label balance for those labels (e.g. sum of all ratios)
- assign image to either test or train
- this assignment won’t change in future, even if future annotations are added 

Optional:
Until an image is assigned to a fixed bucket, we use in the train data for development models

**Comment by Hamzaukw** [comment]
Points from @PeterKczyk 

Data splitting: 
Equal split required such that the classes present in training do not differ very much from testing and production.
Data splitting should be done with respect to data distributions such that we do not miss to train/test for specific classes or even induce data leakage.
 
Label imbalances: 
Some classes may be heavily underrepresented within the dataset leading to unbalanced training which in turn worsens predictive power of our model. 
I would recommend to check for instance occurrences and then evaluate which label set should be used for training the first model while keeping the rest of the data in the back. 
Once a baseline model is established and new annotations are made, we can set a threshold to retrain the model based on the old and new data (at best including the remaining classes).

MLOps: 
I think we should generally move towards MLOps for the overall system. 
At best the predictive model is fed automatically with new data from the annotation frontend when reaching a certain defined threshold. 
Working in a productive environment requires constant reiterations and checks and fallback versions. 
This can be automated via a wide range of tools with different focus.
 Here is a long list: https://github.com/kelvins/awesome-mlops.

**Comment by Hamzaukw** [comment]
@PeterKczyk 

 Label Frequencies:
polyp                          3140
snare                          2886
blood                          2721
stool                          2345
low_quality                    1956
ileocaecalvalve                1826
appendix                       1759
grasper                        1618
outside                        1437
bubbles                        1377
ileum                          1276
wound                          1133
water_jet                      1113
nbi                            941
cleaning                       896
petechia                       628
diverticule                    171
clip                           128
needle                         115
endocuff                       93
inflammation                   59
endoscope_reflection           39
endoscope_retroflection        19
anal_channel                   7


  Unexpected Labels Found:
- endoscope_retroflection
- clip
- endocuff
- anal_channel
- endoscope_reflection
- petechia
- inflammation

**Comment by Hamzaukw** [comment]
### Data Splitting(Avoid data leakage + preserve class distribution):-
#### HybridSplitter (in hybrid_splitter.py):

  - Uses old_examination_id to avoid leakage across splits (group-wise).
  - Stratifies by label distribution (MultiLabelBinarizer) so each split has similar class balance.
    **stratifies:** It means the split algorithm tries to ensure that rare and common labels are distributed similarly between train, val, and test. Not exact stratification like StratifiedKFold, but close approximation for multi-label grouped data.
  - Persistent buckets (via bucket_map.pkl) ensure consistency over multiple runs.


### Label Imbalances
- WeightedRandomSampler (based on inverse label frequency).

**Comment by Hamzaukw** [comment]
Need to test two different videos, look for label **Outside**

**Comment by Maddonix** [comment]
Blocked until we have first production data
